cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/SparkProgramStatusMetricsProvider.java

```diff
@@ -0,0 +1,163 @@
+import com.google.gson.JsonArray;
+import com.google.gson.JsonParser;
+import io.cdap.cdap.api.retry.RetryableException;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.common.service.Retries;
+import io.cdap.cdap.common.service.RetryStrategies;
+import io.cdap.cdap.internal.events.http.SparkApplicationsResponse;
+import io.cdap.cdap.proto.ProgramType;
+import io.cdap.cdap.proto.id.ProgramRunId;
+import io.cdap.common.http.HttpRequest;
+import io.cdap.common.http.HttpRequestConfig;
+import io.cdap.common.http.HttpRequests;
+import io.cdap.common.http.HttpResponse;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.net.URL;
+import java.time.Duration;
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+import java.time.format.DateTimeFormatter;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.StreamSupport;
+import javax.inject.Inject;
+
+/**
+ * Implementation for {@link MetricsProvider}. Retrieves metrics for a Spark program execution.
+ */
+public class SparkProgramStatusMetricsProvider implements MetricsProvider {
+
+  //Pending on actual location
+  private final String sparkBaseUrlConfiguration = "spark.metrics.host";
+  private final String sparkApplicationsEndpoint = "/api/v1/applications";
+  //2 min max for retries strategies expressed in seconds
+  private final long maxTimeRetries = 120;
+
+  private final Logger logger = LoggerFactory.getLogger(SparkProgramStatusMetricsProvider.class);
+  private final JsonParser jsonParser = new JsonParser();
+  private final Gson gson = new Gson();
+
+  private final CConfiguration cConf;
+  private final HttpRequestConfig httpRequestConfig;
+
+  @Inject
+  public SparkProgramStatusMetricsProvider(CConfiguration cConf) {
+    int connectionTimeout = cConf.getInt(Constants.HTTP_CLIENT_CONNECTION_TIMEOUT_MS);
+    int readTimeout = cConf.getInt(Constants.HTTP_CLIENT_READ_TIMEOUT_MS);
+    httpRequestConfig = new HttpRequestConfig(connectionTimeout, readTimeout, false);
+
+    this.cConf = cConf;
+  }
+
+  @Override
+  public ExecutionMetrics retrieveMetrics(ProgramRunId runId) {
+    if (!runId.getType().equals(ProgramType.SPARK)) {
+      return ExecutionMetrics.emptyMetrics();
+    }
+    String runIdStr = runId.getRun();
+    String sparkHistoricBaseURL = cConf.get(sparkBaseUrlConfiguration);
+    String applicationsURL = String.format("%s%s?minEndDate=%s", sparkHistoricBaseURL,
+                                           sparkApplicationsEndpoint, generateMaxTerminationDateParam());
+    return Retries.supplyWithRetries(() -> {
+      ExecutionMetrics metrics;
+      HttpResponse applicationResponse;
+      try {
+        applicationResponse = doGet(applicationsURL);
+      } catch (IOException e) {
+        logger.warn("Error retrieving application response, retrying...", e);
```

Comment:
Any reason to have several small try and catch instead of a large try catch block?